### DISCUSSION TITLE
BUG_Unable to Sign In

### DISCUSSION THREAD

**Discussion by HrBingR** [victorbalssa/abacus#89] [Bug] [answered]
**Describe the bug**
My FireFly III instance is protected and running behind Keycloak for sign in. This is accomplished using the Traefik forwardauth middleware which forwards traffic to KeyCloak for auth, then back to FireFly when authenticated, and passes a HTTP header for authentication.

When trying to authenticate via the app, it redirects to KeyCloak as it should, I sign in, and FireFly asks me to authorize Abacus. When I do, Abacus returns the following error: "Failed to get accessToken, JSON Parse error: Unexpected identifier "Authentication""

Not sure if there's anything that can be done app-side to fix by any chance, but my FireFly instance is behind KeyCloak for added security and for SSO, so I don't really want to remove the forwardauth middleware and revert to username/password authentication for web access.

**To Reproduce**
Type my backend URL into the app, along with the created Oauth Client ID, and tap Sign In.

**Expected behavior**
It should sign in.

**Smartphone (please complete the following information):**
 - Device: iPhone 13 Pro
 - OS: iOS 16.2
 - Version 0.4.0

**Comment by firestrife23** [comment]
Similar issue here, but with Authelia.

Content of Log:

Firefly III v5.7.15 ran into an error: Laravel\Passport\Http\Controllers\AuthorizationController::__construct(): Argument #3 ($guard) must be of type Illuminate\Contracts\Auth\StatefulGuard, FireflyIII\Support\Authentication\RemoteUserGuard given.

The error was of type "TypeError".

The error occurred on/at: Thu, 24 Nov 2022 12:45:29 -0600.

This error occurred in file "/var/www/html/vendor/laravel/passport/src/Http/Controllers/AuthorizationController.php" on line 52 with code 0.

There was no user logged in for this error or no user was detected.

The IP address related to this error is: REDACTED (info)
GET URL is: https://REDACTED/oauth/authorize?client_id=3&client_secret=REDACTED&code_challenge=REDACTED&code_challenge_method=S256&redirect_uri=abacusiosapp%3A%2F%2Fredirect&response_type=code&state=ktdmYsEHGx

The full stacktrace is below. If you think this is a bug in Firefly III, you can forward this message to james@firefly-iii.org. This can help fix the bug you just encountered.

If you prefer, you can also open a new issue on GitHub.

The full stacktrace is below:

```
#0 [internal function]: Laravel\Passport\Http\Controllers\AuthorizationController->__construct()
#1 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(915): ReflectionClass->newInstanceArgs()
#2 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(756): Illuminate\Container\Container->build()
#3 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Application.php(860): Illuminate\Container\Container->resolve()
#4 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(692): Illuminate\Foundation\Application->resolve()
#5 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Application.php(845): Illuminate\Container\Container->make()
#6 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Route.php(274): Illuminate\Foundation\Application->make()
#7 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Route.php(260): Illuminate\Routing\Route->getController()
#8 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Route.php(205): Illuminate\Routing\Route->runController()
#9 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(725): Illuminate\Routing\Route->run()
#10 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(141): Illuminate\Routing\Router->Illuminate\Routing\{closure}()
#11 /var/www/html/vendor/laravel/passport/src/Http/Middleware/CreateFreshApiToken.php(50): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#12 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Laravel\Passport\Http\Middleware\CreateFreshApiToken->handle()
#13 /var/www/html/vendor/laravel/framework/src/Illuminate/Session/Middleware/AuthenticateSession.php(40): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#14 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\Session\Middleware\AuthenticateSession->handle()
#15 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/VerifyCsrfToken.php(78): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#16 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\Foundation\Http\Middleware\VerifyCsrfToken->handle()
#17 /var/www/html/vendor/laravel/framework/src/Illuminate/View/Middleware/ShareErrorsFromSession.php(49): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#18 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\View\Middleware\ShareErrorsFromSession->handle()
#19 /var/www/html/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(121): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#20 /var/www/html/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(64): Illuminate\Session\Middleware\StartSession->handleStatefulRequest()
#21 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\Session\Middleware\StartSession->handle()
#22 /var/www/html/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/AddQueuedCookiesToResponse.php(37): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#23 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse->handle()
#24 /var/www/html/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php(67): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#25 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\Cookie\Middleware\EncryptCookies->handle()
#26 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(116): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#27 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(726): Illuminate\Pipeline\Pipeline->then()
#28 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(703): Illuminate\Routing\Router->runRouteWithinStack()
#29 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(667): Illuminate\Routing\Router->runRoute()
#30 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(656): Illuminate\Routing\Router->dispatchToRoute()
#31 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(190): Illuminate\Routing\Router->dispatch()
#32 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(141): Illuminate\Foundation\Http\Kernel->Illuminate\Foundation\Http\{closure}()
#33 /var/www/html/app/Http/Middleware/InstallationId.php(52): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#34 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): FireflyIII\Http\Middleware\InstallationId->handle()
#35 /var/www/html/vendor/laravel/framework/src/Illuminate/Http/Middleware/TrustProxies.php(39): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#36 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\Http\Middleware\TrustProxies->handle()
#37 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#38 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ConvertEmptyStringsToNull.php(31): Illuminate\Foundation\Http\Middleware\TransformsRequest->handle()
#39 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull->handle()
#40 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#41 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TrimStrings.php(40): Illuminate\Foundation\Http\Middleware\TransformsRequest->handle()
#42 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\Foundation\Http\Middleware\TrimStrings->handle()
#43 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php(27): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#44 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\Foundation\Http\Middleware\ValidatePostSize->handle()
#45 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php(86): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#46 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\Foundation\Http\Middleware\PreventRequestsDuringMaintenance->handle()
#47 /var/www/html/app/Http/Middleware/SecureHeaders.php(51): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#48 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): FireflyIII\Http\Middleware\SecureHeaders->handle()
#49 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(116): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#50 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(165): Illuminate\Pipeline\Pipeline->then()
#51 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(134): Illuminate\Foundation\Http\Kernel->sendRequestThroughRouter()
#52 /var/www/html/public/index.php(75): Illuminate\Foundation\Http\Kernel->handle()
#53 {main}

```
The following headers may also be relevant:

```
- remote-user: REDACTED
- remote-email: REDACTED
- connection: close
- host: REDACTED
- x-forwarded-for: REDACTED
- x-forwarded-host: REDACTED
- x-forwarded-method: GET
- x-forwarded-proto: https
- x-forwarded-server: REDACTED
- x-forwarded-ssl: on
- x-forwarded-uri: /oauth/authorize?code_challenge=REDACTED&redirect_uri=abacusiosapp%3A%2F%2Fredirect&client_id=3&response_type=code&state=ktdmYsEHGx
- x-real-ip: REDACTED
- accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
- user-agent: Mozilla/5.0 (iPhone; CPU iPhone OS 16_1_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.1 Mobile/15E148 Safari/604.1
- accept-language: en-US,en;q=0.9
- accept-encoding: gzip, deflate, br
```

**Comment by victorbalssa** [comment]
Hi @HrBingR & @firestrife23, 
Have you tried to create a Oauth2 client with a secret token ?

**Comment by firestrife23** [comment]
@victorbalssa Yup, no dice.

**Comment by eugene-davis** [comment]
Seeing the same issue with Authentik.

Looks to be due to https://github.com/firefly-iii/firefly-iii/issues/6619
A downgrade to 5.7.14 and I can get past this point but instead get a JSON parse error due to an unrecognized token '<'.

**Comment by iWangJiaxiang** [comment]
Got same issue with latest server and app version

**Comment by iWangJiaxiang** [comment]
> Got same issue with latest server and app version

Found the reason for my issue!

If the url contains `_`，I couln't tap login button, like this `https://firefly_iii.example.com`.

So I just changed the my server url to `https://firefly-iii.example.com`. and everything works perfectly now!

**Comment by HrBingR** [comment]
Still experiencing issues signing in here, though the error has changed:

"Failed to get accessToken, JSON Parse error: Unrecognized token '<'"

**Comment by victorbalssa** [comment]
Do you still experience this issue with the last version `0.6.0` @HrBingR?

**Reply by HrBingR** [reply]
Yes this is still problematic, unfortunately.

**Comment by supermamon** [comment]
I think this is more of a Firefly issue rather that Abacus. I have the same setup as @firestrife23. Firefly dev [mentioned](https://github.com/firefly-iii/firefly-iii/issues/6831#issuecomment-1396464868) that it's something he cannot fix and suggest to using Personal Access Tokens instead. Any change that that will come to Abacus?

**Comment by victorbalssa** [comment]
Good catch @supermamon,
I will implement a 2nd option to log into FireflyIII: #93
I prefer Oauth flow, access token option will be hidden by default.

**Comment by victorbalssa** [comment, marked as answer]
@HrBingR 
@eugene-davis 
@firestrife23 
A new OTA update has been released: 0.6.2 with the Personal Access Token as a authentication option.
Let me know if that solves this issue.

**Reply by supermamon** [reply]
Works pretty well!

**Reply by eugene-davis** [reply]
Also working for me now, thanks for the update!

**Reply by HsuJv** [reply]
@victorbalssa 
Sry for the bothering
I'm using the latest software, which is 0.8.1 at this time, on iOS 16.5.
And I'm trying to use the personal access token to log in to my server(v6.0.20) but get "Failed to validate accessToken, please re-check your token or backend URL."
I'm definitely sure that my URL is correct, and I copy the token from the website.
And if I use the token as a Bearer token and `curl` any API following [this page](https://docs.firefly-iii.org/firefly-iii/api/), it returns successfully. 
When I dig into the server logs I've found no HTTP request before this error so it looks like a local check.
Any suggestions?

**Reply by victorbalssa** [reply]
@HsuJv @nonsavant,
If you are using Android version 9 or more and serving FireflyIII unsecured (http), [android is blocking your call by default](https://developer.android.com/training/articles/security-config#CleartextTrafficPermitted:~:text=Starting%20with%20Android%209%20(API%20level%2028)%2C%20cleartext%20support%20is%20disabled%20by%20default).
I will release a new Abacus version `0.9.0` soon with the `cleartextTrafficPermitted` enabled to support unsecured http calls.
But remember that there is a reason why Android decided to disable clear text traffic, unsecured by default.
Thanks

**Reply by HsuJv** [reply]
Hi @victorbalssa 

Thanks for your reply.

Actually, I'm using iOS via HTTPS, but self-signed. Not sure if this affects.

BTW, I once tried oAuth but also fails. It works w/o nginx reverse-proxy. But similar to #41, w/ nginx involved, it suddenly fails.
Since I need HTTPS and prefer an access token, I decided to spend no time on the oAuth way.

BRs.

**Reply by victorbalssa** [reply]
@HsuJv 

On iOS, It should work /w nginx reverse-proxy and personal access token if no other authentication system interferes the access of `/api/v1/*` on your domain (even with unsecured `http` calls).

Do you have access to the api endpoints from safari?

**Reply by HsuJv** [reply]
@victorbalssa ,

I've got no idea how to access the api endpoints from Safari.
But I can tell that on Safari, the direct access to the web app is successful and can perform all actions such as `create a new bill`.
If I changed the URL to any of the API endpoints it will show unauthenticated.

The nginx is only used to upgrade `http` to `https`, I don't have any other services run on the same port. So there will be no other `/api/vi/*` other than the FireflyIII requires.

What's more, as I mentioned before, it pops 
```
"Failed to validate accessToken, please re-check your token or backend URL."
```
before any of the logs show on the server. It looks more like a sore of local check fails.

BRs


P.S.
nginx conf, while 8080 is a fallback port if CAC fails on 443.
```conf
server {
        listen 8080 ;

        location / {
                proxy_pass http://127.0.0.1:52002;
                proxy_http_version 1.1;
                proxy_set_header Upgrade $http_upgrade;
                proxy_set_header Connection $http_connection;
                proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                proxy_set_header X-Scheme $scheme;
                proxy_set_header Host $host;
                proxy_set_header X-Forwarded-Host $host;
                proxy_set_header X-Forwarded-Proto https; 
                proxy_set_header X-Forwarded-Server $host;
                proxy_set_header Upgrade-Insecure-Requests 1;
                proxy_redirect ~^http://(.+)$ https://$1;
        }
}
```

**Reply by victorbalssa** [reply]
@HsuJv 

Can you try with an OAuth Client _and_ secret and Nginx enabled?

**Reply by HsuJv** [reply]
@victorbalssa 
Tried before.

```
BTW, I once tried oAuth but also fails. It works w/o nginx reverse-proxy. But similar to
https://github.com/victorbalssa/abacus/discussions/41, w/ nginx involved, it suddenly fails.
Since I need HTTPS and prefer an access token, I decided to spend no time on the oAuth way.
```

BRs.

**Reply by HsuJv** [reply]
@victorbalssa 
Sorry I misunderstood your reply

I just tried via postman w/ nginx on,
Both oAuth & personal access token works well
![image](https://github.com/victorbalssa/abacus/assets/14556195/3995b1c5-06e2-4a98-8de5-fd38a307b9cd)

There are three requests while using  postman
```bash
"GET /oauth/authorize?code_challenge=......"
"POST /oauth/authorize HTTP/1.1" 302 2549
"POST /oauth/token HTTP/1.1" 200 2829
```

But when switching to the abacus, it fails at the 2nd
```Bash
"GET /oauth/authorize?code_challenge=......"
"POST /oauth/authorize HTTP/1.1" 302 2615
```


BRs.

**Reply by victorbalssa** [reply]
I can't reproduce the bug unfortunately it will take too much time to re-create an nginx with the same config,
If you want to go deeper with your config, I can recommend `charles-proxy`: https://apps.apple.com/app/charles-proxy/id1134218562?platform=iphone to inspect the response of your first call to Abacus: 
- `POST /oauth/authorize HTTP/1.1" 302 2615`

and I will be able to address any issue on the OAuth impl. from Abacus.

**Reply by HsuJv** [reply]
Hi @victorbalssa ,
Is it okay to give me your email address so that I can send you the packet capture in private since there is some sensitive info?
Also, I can provide my credentials.
BRs.

**Reply by victorbalssa** [reply]
you can send it to github@victorbalssa.ca

**Reply by victorbalssa** [reply]
Answer to this discussion (#89) was that Abacus wasn't working with a **custom installed TLS certificate**.
The quick fix was to just use **HTTP** instead of HTTPS.
